### DISCUSSION TITLE
Integrating with ReferralCandy

### DISCUSSION THREAD

**Discussion by sebpowell** [swellstores/community#575] [Help]
I am considering using Swell over Shopify for a new eCommerce store. I would like to offer a referral programme – e.g. give $5 to every person you refer on their first order, and get $5 store credit / cash when they do. 

There are several solutions out there that provide this sort of functionality – Rivo and Referral Candy appear to be some of the more popular ones. Rivo however, only appears to work with Shopify Premium. Referral Candy, from what I gather, is framework-agnostic.

As far as I understand, the flow would look something like this:

1.) User signs up to Referral Candy and gets their unique code / share link.

2.) Other user clicks on share link and starts shopping. At this point, we need to somehow validate the referral link using Referral Candy's API. If valid, we then need to somehow apply the reward to the cart. In my above example, I mentioned a $5 discount. Looking at Swell's API docs, I presume the way this would need to be done is by creating some sort of generic coupon code (let's call it 'REFER_A_FRIEND') and then automatically apply that to the cart. 

3.) On checkout confirmation, we let Referral Candy know the order has been completed and to send the referrer their reward. 

Has anyone succeeded in integrating Referral Candy with Swell, and if so, is what I described above the right way to go about it? Any advice gratefully received.

**Comment by ericingram** [comment]
@sebpowell I think in your 2nd step, you should use a custom field to indicate the referral code was validated and then apply a "promotion" to the cart using the API.

Once we launch the Swell Apps framework, likely in Q1 2025, you can convert the logic to an app and it would be easier to use and distribute to any account.

**Reply by sebpowell** [reply]
@ericingram thank you! I had a look at the 'promotions' UI, and from what I gather promotions are automatically applied to the cart based on certain conditions (I can't apply a promotion 'manually'). 

Given what you mentioned about custom fields, I assumed that the idea would be to apply the promotion if a product is added to the bag with that custom field included, but I can't see in the UI how you might set that kind of rule up?

**Reply by ericingram** [reply]
@sebpowell you can apply promotions manually via API by passing its ID in the request (from the backend, not storefront API):

```
await swell.post('/carts', {
  ...
  promotion_ids: [referralPromotionId],
});
```

You can set a very high "minimum order subtotal" so that it won't apply automatically. Also note that from the backend API, you would need to pass `$promotions: true` to automatically apply promotions, so if you don't need it work to automatically any way then that's another option
